src/murmur/Messages.cpp

```diff
@@ -2104,3 +2104,32 @@ void Server::msgServerConfig(ServerUser *, MumbleProto::ServerConfig &) {
 
 void Server::msgSuggestConfig(ServerUser *, MumbleProto::SuggestConfig &) {
 }
+
+void Server::msgPluginDataTransmission(ServerUser *sender, MumbleProto::PluginDataTransmission &msg) {
+	// a client's plugin has sent us a message that we shall delegate to its receivers
+
+	if (!msg.has_data() || !msg.has_dataid()) {
+		// messages without data and/or without a data ID can't be used by the clients. Thus we don't even have to send it
+		return;
+	}
+
+	if (!msg.has_sendersession()) {
+		msg.set_sendersession(sender->uiSession);
+	}
+
+	// copy needed data from message in order to be able to remove info about receivers from the message as this doesn't
+	// matter for the client
+	size_t receiverAmount = msg.receiversessions_size();
+	const ::google::protobuf::RepeatedField< ::google::protobuf::uint32 > receiverSessions = msg.receiversessions();
+
+	msg.clear_receiversessions();
+
+	for(int i = 0; static_cast<size_t>(i) < receiverAmount; i++) {
+		ServerUser *receiver = qhUsers.value(receiverSessions.Get(i));
+
+		if (receiver) {
+			// We can simply redirect the message we have received to the clients
+			sendMessage(receiver, msg);
```

Comment:
This needs some sort of rate limiting (at least until we're using QUIC or some other way of having multiple independent data streams), otherwise it trivially allows anyone with a fast connection to DOS every client which has a slower one.
In fact, as is, even a client with a slower connection can DOS a client with a fast connection by just listing them multiple times in the `receiversessions` list, we probably want to forbid that as well.

Reply:
Agreed, but this should be configurable on the server side.
Also, at least my plugin relies on ALL messages being relayed, so the rate limit should first fill a buffer or something, and only when this overflows the messages should be dropped finally. this will help in short spike situations.

Reply:
Good point.

Removing duplicates from the `receiversessions` sounds like a good thing to do.

For rate-limiting: Do you think that the usual rate-limiting is enough? Aka: Only allowing so many actions (including sending messages) in a given time-frame. This could also be configured in the server's config.
If a message gets rate-limited though, it'd be dropped.

Maybe a buffer might be an alternative here. However we shouldn't make this too bug either as otherwise this method could be used for DDos'ing the server... :thinking:

Reply:
For now I decided to introduce a rate-limiter specific to plugin messages. It regularly allows for 4 messages per seconds and bursts of up to 15 messages by default.
I think these defaults should be high enough that regular plugins shouldn't run into the limiter. And if they do, there'll at least be an entry in the server's log about it.

Reply:
Thinking about this though I am wondering whether logging introduces the possibility to DOS the server instead... :thinking: